src/mlir/symbol.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "ast/ast.h"
+#include "mlir/IR/Function.h"
+#include "mlir/IR/Types.h"
+#include "mlir/IR/Value.h"
+
+namespace mlir::verona
+{
+  // The symbol table has all declared symbols (with the original node
+  // and the MLIR ounterpart) in a scope. Creating a new scope makes
+  // all future insertions happen at that level, destroying it pops
+  // the scope out of the stack.
+
+  // New scopes are created by creating a new local variable like:
+  //   SymbolScopeT scope(symbolTable);
+  // The destructor pops the scope automatically.
+
+  // We cannot use LLVM's ADT/ScopedHashTable like MLIR's Toy example because
+  // that coped hash table does not allow redefinition, which is a problem when
+  // declaring variables with a type only and then assigning values later.
+  template<class T>
+  class ScopedTable
+  {
+    using MapTy = std::map<std::string, T>;
+    std::vector<MapTy> stack;
+    int currentScope;
+
+  public:
+    ScopedTable() : currentScope(-1)
+    {
+      pushScope();
+    }
+    ~ScopedTable()
+    {
+      popScope();
+      assert(stack.empty() && currentScope == -1);
+    }
+    bool insert(llvm::StringRef key, T value)
+    {
+      auto& frame = stack[currentScope];
+      if (frame.count(key.str()))
+        return false;
+      auto res = frame.emplace(key, value);
+      return res.second;
+    }
+    T lookup(llvm::StringRef key)
+    {
+      for (int i = currentScope; i >= 0; i--)
+      {
+        auto& frame = stack[i];
+        if (frame.count(key.str()))
+          return frame[key.str()];
+      }
+      // Use inScope for this not to happen
+      llvm_unreachable("Symbol not found");
+    }
+    bool inScope(llvm::StringRef key)
+    {
+      auto frame = stack[currentScope];
```

Comment:
```suggestion
auto frame = stack.back();
```
Pedantic: I prefer predicate methods to start with `is`, as in `isInScope`.